# Assignment of WILD-seq barcodes to cells from 10X sequencing
This pipeline extracts the barcodes from a 10X sequencing run and the matched PCR enrichment sequencing and assigns a WILD-seq barcode to cells

## Prerequisits
**cellranger** https://support.10xgenomics.com/single-cell-gene-expression/software/overview/welcome

**samtools** http://www.htslib.org

**bamtofastq** https://support.10xgenomics.com/docs/bamtofastq

**bowtie** http://bowtie-bio.sourceforge.net/manual.shtml

**umi_tools** https://umi-tools.readthedocs.io/en/latest/

**cutadapt** https://cutadapt.readthedocs.io/en/stable/#

## Custom reference for cellranger

A custom referenece for cellranger alignment was created by adding the sequence of the WILD-seq transcript to the mouse genome using cellranger mkref. The relevant files are available in the WILD-seq/data/custom_pHSW8/ folder

```
cellranger mkref --genome=pHSW8_genome --fasta=combined_genome.fa --genes=genes.gtf
```

## Run cellranger for the 10X sequencing run

Run cellranger using the custom reference for alignment.

```
cellranger count --id=${SampleID} --localcores=20 --localmem=50  --transcriptome=pHSW8_genome --fastqs=${FileDirectory} --sample=${SampleID}
```

## From the bam file generated by cellranger extract all reads mapping to the WILD-seq barcode

```
samtools view -b possorted_genome_bam.bam 4T1_pHSW8:5398-5441 > SITTA11_4T1_pHSW8.bam
```




