# Assignment of WILD-seq barcodes to cells from 10X sequencing
This pipeline extracts the barcodes from a 10X sequencing run and the matched PCR enrichment sequencing and assigns a WILD-seq barcode to cells

## Prerequisits
**cellranger** https://support.10xgenomics.com/single-cell-gene-expression/software/overview/welcome

**samtools** http://www.htslib.org

**bamtofastq** https://support.10xgenomics.com/docs/bamtofastq

**bowtie** http://bowtie-bio.sourceforge.net/manual.shtml

**umi_tools** https://umi-tools.readthedocs.io/en/latest/

**cutadapt** https://cutadapt.readthedocs.io/en/stable/#

## Custom reference for cellranger

A custom referenece for cellranger alignment was created by adding the sequence of the WILD-seq transcript to the mouse genome using cellranger mkref. The fasta file for the plasmid sequence is provided in the WILD-seq/data/custom_pHSW8/ folder and can be used to generate your own custom genome file. The WILD-seq transcript is 4T1_pHSW8:5266-6184.

## Run cellranger for the 10X sequencing run

Run cellranger using the custom reference for alignment.

```
cellranger count --id=[SampleID] --localcores=20 --localmem=50  --transcriptome=pHSW8_genome --fastqs=[Directory] --sample=[SampleID]
```

## Extract reads mapping to the WILD-seq barcode
Extract reads mapping to the WILD-seq barcode from the bam file generated by cellranger

```
samtools view -b possorted_genome_bam.bam 4T1_pHSW8:5398-5441 > SITTA11_4T1_pHSW8.bam #Extracts reads mapping to barcode region from bam file
bamtofastq SITTA11_4T1_pHSW8.bam ./bamtofastq #converts bam file to fastq file
cat bamtofastq*_R2*.fastq.gz > bamtofastq_R2.fastq.gz #If required combine files from different lanes
gunzip bamtofastq_R2.fastq.gz #unzip fastq file for further processing

```

## Extract barcode
Extract only the barcode sequences from the reads and prepare a fasta file 
```
egrep -A 2 -B 1 [ATGC]{12}TGCATCGGTTAACCGATGCA[ATGC]{12} bamtofastq_R2.fastq | sed '/^--$/d' > bamtofastq_R2_filtered.fastq
sed -n '1~4s/^@/>/p;2~4p' bamtofastq_R2_filtered.fastq > bamtofastq_R2_filtered.fa
cat bamtofastq_R2_filtered.fa | sed 's/\(.*\)\([ATCG]\{12\}TGCATCGGTTAACCGATGCA[ATCG]\{12\}\)\(.*\)/\2/g' >  bamtofastq_R2_filtered_cut.fa
```

## Map the WILD-seq barcodes 
The WILD-seq barcode sequences are mapped to the previously generated barcode whitelist bowtie index
```
bowtie -f -v 2 BC_index bamtofastq_R2_filtered_cut.fa > bamtofastq_R2.bowtie
```

## Get the 10X cell barcode and UMI information for the reads
```
samtools view SITTA11_4T1_pHSW8.bam | sed -r -n 's/.*(A00489[:A-Z0-9]*).*(CB:Z:[ACTG]*).*(UB:Z:[ACTG]*).*/\1 \2 \3/p' > SITTA11_4T1_pHSW8.CB.UB.txt
```

## Merge the bowtie output and the 10x cell barcode and UMI information
Here is an example of an R script which will do this.
```
library(dplyr)
library(tidyr)

bowtie <- read.table("bamtofastq_R2.bowtie", sep="\t")
colnames(bowtie) <- c("read","strand","barcode_name","offset","seq","qualities","X", "mismatches")
CB.UB <- read.table("SITTA11_4T1_pHSW8.CB.UB.txt")
colnames(CB.UB) <- c("read", "CBC", "UMI")

bowtie$read <- as.character(bowtie$read)
bowtie$read <- gsub (" 3:N:0:0", "", bowtie$read)

table <- merge(CB.UB, bowtie, by="read")
table <- table[,c("CBC", "UMI", "barcode_name")] 
table <- distinct(table) 

write.table(table, "SITTA11_CBC_table_full.txt", sep="\t", row.names = F, quote = F)
```






